content/blog/enforcing-and-automatically-tagging-aws-resources/index.md

```diff
@@ -0,0 +1,747 @@
+            properties: {
+                requiredTags: {
+                    type: "array",
+                    items: { type: "string" },
+                },
+            },
+        },
+        validateResource: (args, reportViolation) => {
+            const config = args.getConfig<AwsTagsPolicyConfig>();
+            const requiredTags = config.requiredTags;
+            if (requiredTags && isTaggable(args.type)) {
+                const ts = args.props[tags];
+                for (const rt of requiredTags) {
+                    if (!ts || !ts[rt]) {
+                        reportViolation(
+                            `Taggable resource '${args.urn}' is missing required tag '${rt}'`);
+                    }
+                }
+            }
+        },
+    }],
+});
+
+interface AwsTagsPolicyConfig {
+    requiredTags?: string[];
+}
+```
+
+```csharp
+import * as policy from "@pulumi/policy";
+import { isTaggable } from "../lib/taggable";
+
+new policy.PolicyPack("aws-tags-policies", {
+    policies: [{
+        name: "check-required-tags",
+        description: "Ensure required tags are present on all AWS resources.",
+        configSchema: {
+            properties: {
+                requiredTags: {
+                    type: "array",
+                    items: { type: "string" },
+                },
+            },
+        },
+        validateResource: (args, reportViolation) => {
+            const config = args.getConfig<AwsTagsPolicyConfig>();
+            const requiredTags = config.requiredTags;
+            if (requiredTags && isTaggable(args.type)) {
+                const ts = args.props[tags];
+                for (const rt of requiredTags) {
+                    if (!ts || !ts[rt]) {
+                        reportViolation(
+                            `Taggable resource '${args.urn}' is missing required tag '${rt}'`);
+                    }
+                }
+            }
+        },
+    }],
+});
+
+interface AwsTagsPolicyConfig {
+    requiredTags?: string[];
+}
+```
+
+This project defines a _policy pack_ containing a set of _policy rules_, in this case, just one. That rule takes in a configurable set of tags to ensure exist on every taggable AWS resource. If a tag is missing, that is reported as a violation, and the deployment fails.
+
+> This leverages a library that helps to identify taggable AWS resources (see this gist for the full details). Although this policy is written in TypeScript, it can be applied to stacks written in any language.
+
+At this point, we can apply our policy pack to our infrastructure project in two ways: at the CLI or in the SaaS web console.
+
+### Applying Tags Enforcement in the CLI
+
+This policy pack is configurable so that you can enforce whatever tags you want to use without needing to change the pack's code. This makes it more reusable. For the CLI scenario, we will create a `policy-config.json` file that specifies those same three required tags shown above:
```

Comment:
```suggestion
This policy pack is configurable so that you can enforce any tags you want to use without needing to change the pack's code, making it reusable. For the CLI scenario, we will create a `policy-config.json` file that specifies the same three required tags shown above:
```

Reply:
```suggestion
This policy pack is configurable so that you can enforce arbitrary tags without needing to change the pack's code, making it reusable. For the CLI scenario, we will create a `policy-config.json` file that specifies the same three required tags shown above:
```